File: ProgrammingInterviewElements.CSharp.Tests/BitOperations.Tests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Xunit;

namespace ProgrammingInterviewElements.CSharp.Tests
{
    public class BitOperationsTests
    {
        private static readonly Random generator = new Random();

        private static ulong randomUlong()
        {
            var bytes = new byte[sizeof(ulong)];
            generator.NextBytes(bytes);
            return BitConverter.ToUInt64(bytes, 0);
        }

        private static List<ulong> randomUlongs(int quantity)
        {
            quantity = quantity < 1 ? 1 : quantity;
            var result = new List<ulong>(quantity);
            for (int i = 0; i < quantity; ++i)
            {
                result.Add(randomUlong());
            }
            return result;
        }

        #region Problem 5.1 Tests

        private static readonly List<ulong> Evens4Bit =
            new List<ulong> {
                0UL, 3UL, 5UL, 6UL, 9UL, 10UL, 12UL, 15UL
            };

        private static readonly List<ulong> Odds4Bit =
           new List<ulong> {
                1UL, 2UL, 4UL, 7UL, 8UL, 11UL, 13UL, 14UL
            };

        [Fact]
        public void EvenParity4BitNumsTest()
        {
            foreach(ulong x in Evens4Bit)
            {
                Assert.True(x.hasEvenParity());
            }
            foreach(ulong x in Odds4Bit)
            {
                Assert.False(x.hasEvenParity());
            }
        }

        [Fact]
        public void OddParity4BitNumsTest()
        {
            foreach(ulong x in Odds4Bit)
            {
                Assert.True(x.hasOddParity());
            }
            foreach(ulong x in Evens4Bit)
            {
                Assert.False(x.hasOddParity());
            }
        }

        [Fact]
        public void EvenParity4BitNumCollectionTest()
        {
            //Should be true no matter how many even parity nums are combined
            Assert.True(Evens4Bit.hasEvenParity());
            //8 total Odd Parity nums combined, so should be even
            Assert.True(Odds4Bit.hasEvenParity());
        }

        [Fact]
        public void OddParity4BitNumCollection()
        {
            Assert.False(Evens4Bit.hasOddParity());
            Assert.False(Odds4Bit.hasOddParity());
        }

        #endregion

        #region Problem 5.2 Tests

        [Fact]
        public void swapBitsArgOneOverRange()
        {
            ulong x = uint.MaxValue; //0x00000000ffffffff
            Assert.Equal(x.swapBits(64, 0), x);
        }

        [Fact]
        public void swapBitsArgOneUnderRange()
        {
            ulong x = uint.MaxValue; //0x00000000ffffffff
            Assert.Equal(x, x.swapBits(-1, 63));
        }

        [Fact]
        public void swapBitsArgTwoOverRange()
        {
            ulong x = uint.MaxValue; //0x00000000ffffffff
            Assert.Equal(x, x.swapBits(1, 64));
        }

        [Fact]
        public void swapBitsArgTwoUnderRange()
        {
            ulong x = uint.MaxValue; //0x00000000ffffffff
            Assert.Equal(x, x.swapBits(1, -1));
        }

        [Fact]
        public void swapBitsArgsEqual()
        {
            ulong x = uint.MaxValue; //0x00000000ffffffff
            Assert.Equal(x, x.swapBits(0, 0));
        }

        [Fact]
        public void swapBitsIGreaterThanJ()
        {
            ulong x = uint.MaxValue; //0x00000000ffffffff
            ulong expected = (x - 1UL) | (1UL << 63);
            Assert.NotEqual(x.swapBits(63, 0), x);
            Assert.Equal(expected, x.swapBits(63, 0));
        }

        [Fact]
        public void swapBitsILessThanJ()
        {
            ulong x = uint.MaxValue; //0x00000000ffffffff
            ulong expected = (x - 1UL) | (1UL << 63);
            Assert.NotEqual(x.swapBits(0, 63), x);
            Assert.Equal(expected, x.swapBits(0, 63));
        }

        #endregion

        #region Problem 5.3 Tests

        [Fact]
        public void reverseBitsTest()
        {
            var testRuns = 30;
            foreach(ulong x in randomUlongs(testRuns))
            {
                ulong result = x.reverseBits();
                string resultString = Convert.ToString((long)result, 2).PadLeft(64, '0');
                string xString = Convert.ToString((long)x, 2).PadLeft(64, '0');               
                Assert.Equal(xString.Reverse(), resultString);
            }
        }

        #endregion

        #region Problem 5.4 Tests

        [Fact]
        public void nearestEqualWeightWorksForZero()
        {
            ulong x = 0UL;
            Assert.Equal(x, x.nearestEqualWeight());
        }

        [Fact]
        public static void nearestEqualWeightWorksForAllOnes()
        {
            ulong x = ulong.MaxValue;
            Assert.Equal(x, x.nearestEqualWeight());
        }

        private static ulong weight(ulong x)
        {
            ulong result = 0;
            for(int i = 0; i < 64; ++i)
            {
                result += x & 1UL;
                x >>= 1;
            }
            return result;
        }

        [Fact]
        public static void nearestEqualWeightFindsEqualWeightAnswer()
        {
            foreach (ulong x in randomUlongs(30))
            {
                Assert.Equal(weight(x), weight(x.nearestEqualWeight()));
            }
        }

        [Fact]
        public static void nearestEqualWeightFindsCloseValue()
        {
            foreach(ulong x in randomUlongs(30))
            {
                ulong result = x.nearestEqualWeight();
                ulong checkedDifference = result > x ? result - x : x - result;
                Assert.True(checkedDifference <= Math.Min(x, result));
            }
        }
        
        // Some specific cases, as don't know how to test in general
        // without repeating solution

        [Fact]
        public static void nearestEqualWeightTo12is10()
        {
            ulong x = 12UL;
            ulong expected = 10UL;
            Assert.Equal(expected, x.nearestEqualWeight());
        }

        [Fact]
        public static void nearestEqualWeightTo8is4()
        {
            ulong x = 8UL;
            ulong expected = 4UL;
            Assert.Equal(expected, x.nearestEqualWeight());
        }

        [Fact]
        public static void nearestEqualWeightToOneZeroAllOnesIsOneOneZeroAllOnes()
        {
            ulong expected = ulong.MaxValue & (1UL << 61);
            ulong x = ulong.MaxValue & (1UL << 62);
            Assert.Equal(expected, x.nearestEqualWeight());
        }

        [Fact]
        public static void nearestEqualWeightToZeroAllOnesIsOneZeroAllOnes()
        {
            ulong expected = ulong.MaxValue & (1UL << 62);
            ulong x = ulong.MaxValue & (1UL << 63);
            Assert.Equal(expected, x.nearestEqualWeight());
        }

        #endregion
    }
}

File: FSharpTests/BitOperations.Tests.fs
module BitOperations.Tests

open Xunit
open BitOperations
open System
open System.Collections.Generic

let Evens4Bit = Seq.ofList [0UL;3UL;5UL;6UL;9UL;10UL;12UL;15UL]
let Odds4Bit = Seq.ofList [1UL;2UL;4UL;7UL;8UL;11UL;13UL;14UL]
let Odds = Seq.skip 1 Odds4Bit //an odd number of odds
let isEven x = x % 2 = 0
let isOdd x = not <| isEven x

// Problem 5.1 Tests
// Calculate parity of a sequence of numbers

[<Fact>]
let hasEvenParityPassesForEvens() =
    Evens4Bit |> Seq.iter (Assert.True << hasEvenParity)

[<Fact>]
let hasEvenParityFailsForOdds() =
    Odds4Bit |> Seq.iter (Assert.False << hasEvenParity)  

[<Fact>]
let hasOddParityPassesForOdds() =
    Odds4Bit |> Seq.iter (Assert.True << hasOddParity)

[<Fact>]
let hasOddParityFailsForEvens() =
    Evens4Bit |> Seq.iter (Assert.False << hasOddParity)

[<Fact>]
let hasEvenParitySeqPassesForEvens() =
    Assert.True <| Seq.hasEvenParity Evens4Bit

[<Fact>]
let hasEvenParitySeqPassesForEvenNumberOfOdds() =
    Assert.True <| (Seq.length Odds4Bit |> isEven)
    Assert.True <| Seq.hasEvenParity Odds4Bit

[<Fact>]    
let hasEvenParitySeqFailsForOddNumberOfOdds() =
    Assert.True <| (Seq.length Odds |> isOdd)
    Assert.False <| Seq.hasEvenParity Odds

[<Fact>]
let hasOddParitySeqPassesForOddNumberOfOdds() =
    Assert.True <| (Seq.length Odds |> isOdd)
    Assert.True <| Seq.hasOddParity Odds

[<Fact>]
let hasOddParitySeqFailsForEvenNumberOfOdds() =
    Assert.True <| (Seq.length Odds4Bit |> isEven)
    Assert.False <| Seq.hasOddParity Odds4Bit

[<Fact>]
let hasOddParitySeqFailsForEvens() =
    Assert.False <| Seq.hasOddParity Evens4Bit
    
// Problem 5.2 Tests
// Swap bits by index

[<Fact>]
let swapBitsUlongReturnsIntialValueWhenBitPositionsEqual =
    let x = uint64 UInt32.MaxValue
    Assert.Equal(x, x |> swapBits (BitPosition 0) (BitPosition 0))

[<Fact>]
let swapBitsUlongSwapsWhenArg1GreaterThanArg2 =
    let x = uint64 UInt32.MaxValue
    let expected = (x - 1UL) ||| (1UL <<< 63)
    Assert.Equal(expected, x |> swapBits (BitPosition 63) (BitPosition 0))

[<Fact>]
let swapBitsULongSwapsWhenArg1LessThanArg2 =
    let x = uint64 UInt32.MaxValue
    let expected = (x - 1UL) ||| (1UL <<< 63)
    Assert.Equal(expected, x |> swapBits (BitPosition 0) (BitPosition 63))

// Problem 5.3 Tests
// Reverse bits in a number

type UInt64 with
    static member random =
        let r = Random()
        let bytes = Array.create sizeof<ulong> 0uy
        let internalRandom() =
            r.NextBytes bytes
            BitConverter.ToUInt64(bytes, 0)
        internalRandom
    static member randoms count =
        List.init count (fun _ -> ulong.random())   
                 
[<Fact>]
let reverseBitsReversesBits() = 
    let test x =
        let result = reverseBits x
        let resultString = Convert.ToString(int64 result,2).PadLeft(64,'0')
        let givenString = Convert.ToString(int64 x,2).PadLeft(64,'0')
        let givenStringReversed = String(Array.rev(givenString.ToCharArray()))
        Assert.Equal(givenStringReversed, resultString)
    List.iter test <| UInt64.randoms 30

// Problem 5.4 Tests
// Find nearest number with equal weight (equal number of 1's)

[<Fact>]
let nearestEqualWeightWorksForZero() =
    Assert.Equal(0UL, nearestEqualWeight 0UL)

[<Fact>]
let nearestEqualWeightWorksForAllOnes() =
    Assert.Equal(UInt64.MaxValue, nearestEqualWeight UInt64.MaxValue)

let weight x =
    let folder acc i = match getBit (BitPosition i) x with
                       | Zero -> acc
                       | One -> acc + 1
    {1..64} |> Seq.fold folder 0

[<Fact>]
let nearestEqualWeightFindsEqualWeightAnswer() =
    let test x = Assert.Equal(weight x, weight <| nearestEqualWeight x)
    List.iter test <| UInt64.randoms 30 

[<Fact>]
let nearestEqualWeightFindsCloseValue() =
    let checkedDifference x y = 
        if x > y then x - y else y - x
    let test x =
        let testVal = nearestEqualWeight x
        Assert.True(checkedDifference x testVal <= Math.Min(x,testVal))
    List.iter test <| UInt64.randoms 30

// Not sure how to really test this in general without building the solution
// Brute force test would take too long 
// So, test a few cases
let testNearestEqualWeightVals expected testValue =
    Assert.Equal(expected, nearestEqualWeight testValue)

[<Fact>]
let nearestEqualWeightTo12is10() =
    testNearestEqualWeightVals 10UL 12UL

[<Fact>]
let nearestEqualWeightTo8is4() =
    testNearestEqualWeightVals 4UL 8UL

[<Fact>]
let nearestEqualWeightToOneZeroAllOnesIsOneOneZeroAllOnes() =
    let expected = clearBit (BitPosition 61) UInt64.MaxValue
    let testVal = clearBit (BitPosition 62) UInt64.MaxValue
    testNearestEqualWeightVals expected testVal

[<Fact>]
let nearestEqualWeightToZeroAllOnesIsOneZeroAllOnes() =
    let expected = clearBit (BitPosition 62) UInt64.MaxValue
    let testVal = clearBit (BitPosition 63) UInt64.MaxValue
    testNearestEqualWeightVals expected testVal

// Problem 5.5 Tests
// Multiply 2 unsigned numbers without + or *

[<Fact>]
let bitAddAdds() =
    let test (i,j) = 
        Assert.Equal(i + j, bitAdd i j)
    let sample = List.zip (UInt64.randoms 30) (UInt64.randoms 30)
    List.iter test sample

[<Fact>]
let bitMultMultiplies() =
    let test (i,j) = 
        Assert.Equal(i * j, bitMult i j)
    let sample = List.zip (UInt64.randoms 30) (UInt64.randoms 30)
    List.iter test sample

// Problem 5.6 Tests
// Divide 2 unsigned numbers with +, -, shift

[<Fact>]
let bitDivDivides() =
    let test (i,j) =
        let j' = if j = 0UL then 1UL else j
        Assert.Equal(Some (i / j'), tryBitDiv i j')                            
    let sample = List.zip (UInt64.randoms 30) (UInt64.randoms 30)
    List.iter test sample

[<Fact>]
let bitDivReturnsNoneOnDivideByZero() = 
    let test i =
        Assert.Equal(None, tryBitDiv i 0UL)
    let sample = UInt64.randoms 30
    List.iter test sample